NetworkPkg/HttpBootDxe/HttpBootImpl.c

```diff
@@ -370,13 +374,42 @@ HttpBootGetBootFileCaller (
         //
         // Load the boot file into Buffer
         //
-        Status = HttpBootGetBootFile (
-                   Private,
-                   FALSE,
-                   BufferSize,
-                   Buffer,
-                   ImageType
-                   );
+        for (Retries = 1; ; Retries++) {
+          Status = HttpBootGetBootFile (
+                     Private,
+                     FALSE,
+                     BufferSize,
+                     Buffer,
+                     ImageType
+                     );
+          if (!EFI_ERROR (Status) ||
+              (Status != EFI_NOT_READY) ||
+              (Retries >= PcdGet32 (PcdMaxHttpResumeRetries)))
+          {
+            break;
+          }
+
+          //
+          // HttpBootGetBootFile returned EFI_NOT_READY, we may attempt to resume
+          // the interrupted download.
+          //
+
+          Private->HttpCreated = FALSE;
+          HttpIoDestroyIo (&Private->HttpIo);
+          Status = HttpBootCreateHttpIo (Private);
+          if (EFI_ERROR (Status)) {
+            break;
+          }
+
+          DEBUG ((DEBUG_WARN | DEBUG_INFO, "HttpBootGetBootFileCaller: NBP file download interrupted, will try to resume the operation.\n"));
+          gBS->Stall (1000 * 1000 * PcdGet32 (PcdHttpDelayBetweenResumeRetries));
+        }
+
+        if (Status == EFI_NOT_READY) {

```

Comment:
In my implementation, only when function HttpBootGetBootFile returns EFI_NOT_READY the download may be tryied again. Other types of errors it should not attempt to resume the download.

Reply:
Maybe instead of changing the error to EFI_NOT_READY  when EFI_TIMEOUT or EFI_DEVICE_ERROR happens inside HttpBootGetBootFile, to indicate that operation may be retried, I could revert this and modifify the code you pointed out to check for EFI_TIMEOUT or EFI_DEVICE_ERROR  to try again. This should improve the case when HttpIoSendRequest time out or first call to HttpIoRecvResponse to get headers also time out. 

What do you think about?